src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs

```diff
@@ -33,9 +36,18 @@ public sealed partial class JsonObject : JsonNode
         /// <param name="options">Options to control the behavior.</param>
         public JsonObject(IEnumerable<KeyValuePair<string, JsonNode?>> properties, JsonNodeOptions? options = null) : this(options)
         {
-            if (properties is ICollection<KeyValuePair<string, JsonNode?>> propertiesCollection)
+#if NET6_0_OR_GREATER
+            if (properties.TryGetNonEnumeratedCount(out int initialCapacity))
             {
-                _initialCapacity = propertiesCollection.Count;
+#else
+            if (properties is ICollection<KeyValuePair<string, JsonNode?>> or IReadOnlyCollection<KeyValuePair<string, JsonNode?>>)
+            {
+                int initialCapacity =
+                    (properties as ICollection<KeyValuePair<string, JsonNode?>>)?.Count // TODO: can I avoid the redundant type checks?
+                    ?? (properties as IReadOnlyCollection<KeyValuePair<string, JsonNode?>>)?.Count
+                    ?? throw null; // should be unreachable, amirite??
```

Comment:
Inlining a `throw` statement inside the method body impacts performance and is something we deliberately avoid in performance-sensitive methods. Just checking for `ICollection<T>` should suffice in my opinion, almost all collection types, including read-only collection types do implement it.

Reply:
Fair point about performance! I was hoping to be able to avoid the double-type-check business (and so avoid the need for that last branch). Looks like if I go back to just checking for `ICollection`, I won't need that anyway.. sorry I think maybe I misunderstood your last PR comment there.

For my own learning, would there be a way to avoid the double type checks here while still checking for both `ICollection` _and_ `IReadOnlyCollection`?

Reply:
Not really. It's for this reason that `TryGetNonEnumeratedCount` does not check for `IReadOnlyCollection` to avoid the added perf penalty.